File: source_code/python/bfstree.py
from collections import deque
from node import node

def bfstree(s0):
    # s0 = initial state

    # create the initial node
    n0 = node(s0, None, None, 0, 0) 

    # initizlize #visited nodes
    nvisited = 0
    
    # initialize the frontier list
    frontier = deque([n0])

    while True:
        # the search fails when the frontier is empty
        if not frontier:
            return (None, nvisited)
        else:
            # get one node from the frontier
            n = frontier.popleft()
            # count the number of visited nodes
            nvisited+=1
            # check if the state in n is a goal
            if n.state.isgoal():
                return (n, nvisited)
            else:
                # generate successor states
                S = n.state.successors()
                # create new nodes and add to the frontier
                for (s, a, c) in S:
                    p = node(s, n, a, n.cost+c, n.depth+1)
                    frontier.append(p)

File: source_code/java/BFSTree.java
import java.util.ArrayList;
import java.util.List;

public class BFSTree {
  public int nvisited;

  public Node search(State s0) {
    Node n0 = new Node(s0, null, null, 0, 0);
    nvisited = 0;
    ArrayList<Node> frontier = new ArrayList<Node>();
    frontier.add(n0);
    while(true) {
      if (frontier.isEmpty()) {
        return null;
      }
      else {
        Node n = frontier.remove(0);
        nvisited++;
        if (n.state.isGoal()) {
          return n;
        }
        else {
          List<Successor> succs = n.state.successors();
          for(Successor s : succs) {
            Node p = new Node(s.state, n, s.action,
                n.cost+s.cost, n.depth+1);
            frontier.add(p);
          }
        }
      }
    }
  }
}
